merkletree/tree.go

```diff
@@ -0,0 +1,219 @@
+	// Root returns the root node of the tree
+	Root() *Node
+	// Leafs returns all the leaf nodes in the tree
+	Leafs() []Node
+	// ConstructProof constructs a Merkle proof of the subtree (or leaf) at level lvl with index idx.
+	// level 0 is the root and index 0 is the left-most node in a level.
+	ConstructProof(lvl int, idx int) (MerkleProof, error)
+	// ConstructBatchedProof constructs a batched Merkle proof of the nodes from and including leftLvl, leftIdx, to and including rightLvl, rightIdx.
+	// That is, if leftLvl, or rightLvl, is not the leaf-level, then the proof is of the entire subtree from leftLvl at leftIdx to rightLvl at rightIdx
+	// Level 0 is the root and index 0 is the left-most node in a level.
+	ConstructBatchedProof(leftLvl int, leftIdx int, rightLvl int, rightIdx int) (BatchedMerkleProof, error)
+	// ValidateFromLeafs checks that the Merkle tree is correctly constructed based on all the leaf data
+	ValidateFromLeafs(leafData [][]byte) bool
+	// Validate checks that the Merkle tree is correctly constructed, based on the internal nodes
+	Validate() bool
+}
+
+type data struct {
+	// nodes start from root and go down left-to-right
+	// thus len(nodes[0]) = 1, len(nodes[1]) = 2, etc.
+	nodes [][]Node
+}
+
+type Node struct {
+	data [digestBytes]byte
+}
+
+// NewBareTree allocates that memory needed to construct a tree with a specific amount of leafs
+func NewBareTree(leafs int) MerkleTree {
+	var tree data
+	tree.nodes = make([][]Node, 1+util.Log2Ceil(leafs))
+	for i := 0; i <= util.Log2Ceil(leafs); i++ {
+		tree.nodes[i] = make([]Node, 1<<i)
+	}
+	return tree
+}
+
+// GrowTree constructs a Merkle from a list of leafData, the data of a given leaf is represented as a byte slice
+func GrowTree(leafData [][]byte) (MerkleTree, error) {
+	var tree MerkleTree
+	if leafData == nil || len(leafData) == 0 {
+		return tree, errors.New("empty input")
+	}
+	leafLevel := hashList(leafData)
+	return growTreeHashedLeafs(leafLevel), nil
+}
+
+// growTreeHashedLeafs constructs a tree from leafs nodes, i.e. leaf data that has been hashed to construct a Node
+func growTreeHashedLeafs(leafs []Node) MerkleTree {
+	tree := NewBareTree(len(leafs))
+	// Set the leaf nodes
+	tree.(data).nodes[util.Log2Ceil(len(leafs))] = leafs
+	preLevel := leafs
+	// Construct the Merkle tree bottom-up, starting from the leafs
+	// Note the -1 due to 0-indexing the root level
+	for level := util.Log2Ceil(len(leafs)) - 1; level >= 0; level-- {
+		currentLevel := make([]Node, util.Ceil(len(preLevel), 2))
+		// Traverse the level left to right
+		for i := 0; i+1 < len(preLevel); i = i + 2 {
+			currentLevel[i/2] = *computeNode(&preLevel[i], &preLevel[i+1])
+		}
+		// Handle the edge case where the tree is not complete, i.e. there is an odd number of leafs
+		// This is done by hashing the content of the node and letting it be its own parent
+		if len(preLevel)%2 == 1 {
+			currentLevel[util.Ceil(len(preLevel), 2)-1] = *truncatedHash(preLevel[len(preLevel)-1].data[:])
+		}
```

Comment:
Ahh, this answers my question. This is different from how CommP/D commitment works. We zero pad up to the power of two size. There is an optimization which can be made to make it O(1) operation but we can skip that for now.

FYI https://github.com/filecoin-project/go-commp-utils/blob/master/zerocomm/zerocomm.go